File: js/packages/core/src/integration.ts
import type { IdentifyEvent, IntegrationResult, PageEvent, TrackEvent } from "./types";

/** Optional callbacks fired before an integration handles a call. */
export interface IntegrationHooks {
  onInit?: () => void;
  onIdentify?: (event: IdentifyEvent) => void;
  onTrack?: (event: TrackEvent) => void;
  onPage?: (event: PageEvent) => void;
  onReset?: () => void;
}

/**
 * Base class every integration extends. Override what the vendor supports and
 * call `super` first so the hooks run, exactly like the Flutter package.
 */
export class Integration {
  /** Name used in results and logs; defaults to the class name. */
  readonly name: string;
  protected readonly hooks: IntegrationHooks;

  constructor(hooks: IntegrationHooks = {}, name?: string) {
    this.hooks = hooks;
    this.name = name ?? new.target.name ?? "Integration";
  }

  async init(): Promise<void> {
    this.hooks.onInit?.();
  }

  async identify(event: IdentifyEvent): Promise<void> {
    this.hooks.onIdentify?.(event);
  }

  async track(event: TrackEvent): Promise<void> {
    this.hooks.onTrack?.(event);
  }

  async page(event: PageEvent): Promise<void> {
    this.hooks.onPage?.(event);
  }

  async reset(): Promise<void> {
    this.hooks.onReset?.();
  }

  /** Sends anything buffered. No-op by default. */
  async flush(): Promise<void> {}

  /** Flushes and releases timers or sockets. No-op by default. */
  async shutdown(): Promise<void> {}
}

/** Runs one integration call in isolation and reports the outcome. */
export async function safeExecute(integration: Integration, action: () => Promise<void> | void): Promise<IntegrationResult> {
  try {
    await action();
    return { integration: integration.name, success: true, timestamp: new Date() };
  } catch (error) {
    return { integration: integration.name, success: false, error, timestamp: new Date() };
  }
}

File: ruby/iforevents/lib/iforevents/integration.rb
# frozen_string_literal: true

module Iforevents
  # Base class every integration extends; mirrors the Flutter +Integration+.
  # Override what the vendor supports and call +super+ first so hooks run.
  class Integration
    attr_reader :name

    def initialize(name: nil, on_init: nil, on_identify: nil, on_track: nil, on_page: nil, on_reset: nil)
      @name = name || self.class.name.split("::").last
      @on_init = on_init
      @on_identify = on_identify
      @on_track = on_track
      @on_page = on_page
      @on_reset = on_reset
    end

    def init
      @on_init&.call
    end

    def identify(event)
      @on_identify&.call(event)
    end

    def track(event)
      @on_track&.call(event)
    end

    def page(event)
      @on_page&.call(event)
    end

    def reset
      @on_reset&.call
    end

    # Sends anything buffered. No-op by default.
    def flush; end

    # Flushes and releases resources. No-op by default.
    def shutdown; end
  end

  # Runs one integration call in isolation and reports the outcome.
  def self.safe_execute(integration)
    yield
    IntegrationResult.new(integration: integration.name, success: true, error: nil, timestamp: Time.now.utc)
  rescue StandardError => e
    IntegrationResult.new(integration: integration.name, success: false, error: e, timestamp: Time.now.utc)
  end
end
